src/crystal/system/win32/event_loop_iocp.cr

```diff
@@ -35,35 +39,55 @@ class Crystal::Iocp::EventLoop < Crystal::EventLoop
     iocp
   end
 
+  def run_once : Nil
+    run(blocking: true)
+  end
+
   # Runs the event loop and enqueues the fiber for the next upcoming event or
   # completion.
-  def run_once : Nil
+  def run(blocking : Bool) : Bool
     # Pull the next upcoming event from the event queue. This determines the
     # timeout for waiting on the completion port.
     # OPTIMIZE: Implement @queue as a priority queue in order to avoid this
     # explicit search for the lowest value and dequeue more efficient.
     next_event = @queue.min_by?(&.wake_at)
 
-    unless next_event
-      Crystal::System.print_error "Warning: No runnables in scheduler. Exiting program.\n"
-      ::exit
-    end
+    # no registered events: nothing to wait for
+    return false unless next_event
 
     now = Time.monotonic
 
     if next_event.wake_at > now
-      wait_time = next_event.wake_at - now
-      # There is no event ready to wake. So we wait for completions with a
-      # timeout for the next event wake time.
+      # There is no event ready to wake. We wait for completions until the next
+      # event wake time, unless nonblocking or already interrupted (timeout
+      # immediately).
+      if blocking
+        @lock.sync do
+          if @interrupted.get(:acquire)
+            blocking = false
+          else
+            # memorize the blocked thread (so we can alert it)
+            @blocked_thread.set(Thread.current, :release)
```

Comment:
**question:** I reckon this means the scheduler must guarantee to never run the event loop on the same `EventLoop` instance in different threads simultaneously?

Reply:
I assume as much; even if the event loop is shared across N threads, I assume only one thread needs to run/wait on it at a time.

Do you see benefits to running/waiting on the same event loop in different threads?

Reply:
I'm not sure. But it's a possible scenario. There could be use cases where it makes sense. We intent to allow custom implementations of execution contexts (and therefore schedulers), so this could actually come up.
Until there's some actual demand, we probably don't need to do anything to support multi-threaded parallel runs.
But we should at least mention this limitation in the documentation.

Reply:
I might have a scenario in mind, to simplify park/wakeup. For example epoll for a fd with EPOLLET (edge triggered) will only wakeup one waiting thread (nice). That would make `#interrupt` pointless.

But this all depends on how the other systems are working (kqueue, io_uring, IOCP).

Reply:
Yes, we need to check which system selectors allow this.

But I think this should work with most of them. Pretty sure about IOCP and I suppose any completion-based mechanism because there an activation is associated with a specific queued action being completed.